### PR TITLE
[DependencyInjection] deprecate access to private shared services.

### PR DESCRIPTION
**Pull request by hhamon**
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | yes |
| Tests pass? | yes |
| Fixed tickets | #19117 |
| License | MIT |
| Doc PR | ~ |

### CODE REVIEW

**Comment by hhamon** [comment]
Looks like fabbot gives a false positive!

**Comment by fabpot** [comment]
But tests are truly broken :)

File: UPGRADE-4.0.md [hunk 1]
```diff
@@ -18,6 +18,16 @@ DependencyInjection
...
+ * Resetting a private shared service from outside of the container with the
+   `set()` method of the `Container` class is no longer supported and will raise
+   an exception.
+
+ * Checking the existence of a private shared service with the `has()` method of
+   the `Container` class is no longer supported and will raise an exception.
+
+ * Requesting a private shared service with the `get()` method of the `Container`
+   class is no longer supported and will raise an exception.
+
```
Comment (fabpot) [inline UPGRADE-4.0.md]: You also need to add a note in the component CHANGELOG file.

File: UPGRADE-4.0.md [hunk 2]
```diff
@@ -18,6 +18,16 @@ DependencyInjection
 
  * Using unsupported options to configure service aliases raises an exception.
 
+ * Resetting a private shared service from outside of the container with the
+   `set()` method of the `Container` class is no longer supported and will raise
+   an exception.
```
Comment (fabpot) [inline UPGRADE-4.0.md]: I think we need to be more precise here. It won't raise any new exception, it will just tell the dev that the service does not exist.

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php [hunk 3]
```diff
@@ -606,12 +607,11 @@ private function addService($id, $definition)
         }
 
         if (!$definition->isPublic()) {
+            $args[] = '$triggerDeprecation = true';
             $doc .= <<<'EOF'
 
      *
      * This service is private.
-     * If you want to be able to request this service from the container directly,
-     * make it public, otherwise you might end up with broken code.
```
Comment (fabpot) [inline src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: This is still valid.

**Comment by fabpot** [comment]
I would have implemented this feature in a different way by renaming internally private services so that their original name is not available from the container (this is what we are doing for anonymous services for instance).

**Comment by xabbuh** [comment]
I think it's a good idea to switch to generated ids in 4.0. For now I like this approach as it allows users to fix their applications more smoothly.

File: src/Symfony/Component/DependencyInjection/Container.php [hunk 4]
```diff
@@ -198,6 +203,10 @@ public function has($id)
             if (--$i && $id !== $lcId = strtolower($id)) {
                 $id = $lcId;
             } else {
+                if (isset($this->privates[$id])) {
+                    @trigger_error(sprintf('Checking for the existence of the private service "%s" is deprecated since Symfony 3.2 and won\'t be supported anymore in Symfony 4.0. An exception will be thrown instead in Symfony 4.0.', $id), E_USER_DEPRECATED);
```
Comment (javiereguiluz) [inline src/Symfony/Component/DependencyInjection/Container.php]: Why throw an exception in this case? If you check a private service for existence, the method should return `false`, right?

Comment (fabpot) [inline src/Symfony/Component/DependencyInjection/Container.php]: Yes

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php [hunk 5]
```diff
@@ -606,12 +607,11 @@ private function addService($id, $definition)
         }
 
         if (!$definition->isPublic()) {
+            $args[] = '$triggerDeprecation = true';
```
Comment (Ener-Getick) [inline src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: Should be removed as the deprecation is not trigerred at this level.

File: src/Symfony/Component/DependencyInjection/Container.php (hunk 4 above)
Comment (hhamon) [inline src/Symfony/Component/DependencyInjection/Container.php]: Yes it makes more sense!

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php (hunk 5 above)
Comment (hhamon) [inline src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: True!

**Comment by fabpot** [comment]
@xabbuh Fair enough, let's keep the current approach for now.

File: UPGRADE-4.0.md [hunk 6]
```diff
@@ -18,6 +18,16 @@ DependencyInjection
...
+
+ * Checking the existence of a private shared service with the `has()` method of
```
Comment (sstok) [inline UPGRADE-4.0.md]: Is this correct? Throwing an exception for a method that checks existence is not good for DX IMO.

**Comment by hhamon** [comment]
I'll update the code later today to reflect your comments.

File: src/Symfony/Component/DependencyInjection/Container.php [hunk 7]
```diff
@@ -176,6 +177,10 @@ public function set($id, $service)
         if (null === $service) {
             unset($this->services[$id]);
         }
+
+        if (isset($this->privates[$id])) {
+            @trigger_error(sprintf('Resetting the private service "%s" from outside of the container is deprecated since Symfony 3.2 and won\'t be supported anymore in Symfony 4.0. An exception will be thrown instead in Symfony 4.0.', $id), E_USER_DEPRECATED);
```
Comment (sstok) [inline src/Symfony/Component/DependencyInjection/Container.php]: `. and` is part of new sentence part and should start with an uppercase 👍

Comment (hhamon) [inline src/Symfony/Component/DependencyInjection/Container.php]: Where is that?

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php [hunk 8]
```diff
@@ -889,6 +889,36 @@ private function addMethodMap()
     }
 
     /**
+     * Adds the $privates property definition.
```
Comment (sstok) [inline src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: To be consistent with other comments this should not contain the `$` sign.

File: src/Symfony/Component/DependencyInjection/Tests/ContainerTest.php [hunk 9]
```diff
@@ -306,11 +306,39 @@ public function testThatCloningIsNotSupported()
         $this->assertFalse($class->isCloneable());
         $this->assertTrue($clone->isPrivate());
     }
+
+    /** @group legacy */
+    public function testUnsetInternalPrivateServiceIsDeprecated()
+    {
+        $c = new ProjectServiceContainer();
+        $c->set('internal', null);
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Tests/ContainerTest.php]: A test without any assertion is wrong.

You should assert that the deprecation is triggered instead

Comment (hhamon) [inline src/Symfony/Component/DependencyInjection/Tests/ContainerTest.php]: Do we have an assertion method for PHPUnit to check for a triggered notice?

**Comment by stof** [comment]
@fabpot generated ids are a good idea for 4.0 indeed, but this cannot be done for deprecations. If we want to switch to generated ids already, we would have to build another feature to to store a mapping between shared private service ids and their generated ids, to retrieve them this way and trigger a deprecation, which would again look like this approach.

File: src/Symfony/Component/DependencyInjection/Tests/ContainerTest.php (hunk 9 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Tests/ContainerTest.php]: See #18880

Comment (hhamon) [inline src/Symfony/Component/DependencyInjection/Tests/ContainerTest.php]: Thanks @nicolas-grekas. I'll update the PR this evening.

**Comment by hhamon** [comment]
I don't understand why the tests suite fails on PHP 5.5 and HHVM. The failing test in the ProxyManager Bridge passes on my machine with PHP 5.5.36...

File: src/Symfony/Component/DependencyInjection/Container.php [hunk 10]
```diff
@@ -268,6 +281,9 @@ public function get($id, $invalidBehavior = self::EXCEPTION_ON_INVALID_REFERENCE
 
                 return;
             }
+            if (isset($this->privates[$id])) {
+                @trigger_error(sprintf('Requesting the "%s" private service is deprecated since Symfony 3.2 and won\'t be supported anymore in Symfony 4.0. An exception will be thrown instead in Symfony 4.0.', $id), E_USER_DEPRECATED);
```
Comment (fabpot) [inline src/Symfony/Component/DependencyInjection/Container.php]: I won't necessarily throw an exception as it depends on the `$invalidBehavior` value.

File: src/Symfony/Component/DependencyInjection/Container.php [hunk 11]
```diff
@@ -176,6 +177,14 @@ public function set($id, $service)
         if (null === $service) {
             unset($this->services[$id]);
         }
+
+        if (isset($this->privates[$id])) {
+            if (null === $service) {
+                @trigger_error(sprintf('Unsetting the "%s" private service from outside of the container is deprecated since Symfony 3.2 and won\'t be supported anymore in Symfony 4.0. An exception will be thrown instead in Symfony 4.0.', $id), E_USER_DEPRECATED);
+            } else {
+                @trigger_error(sprintf('Resetting the "%s" private service from outside of the container is deprecated since Symfony 3.2 and won\'t be supported anymore in Symfony 4.0. An exception will be thrown instead in Symfony 4.0.', $id), E_USER_DEPRECATED);
+            }
```
Comment (fabpot) [inline src/Symfony/Component/DependencyInjection/Container.php]: If we don't keep the `private` var in 4.0 but replace it with random names, setting here will work as the container will just define a new service, which will probably confuse people.

File: src/Symfony/Component/DependencyInjection/Container.php (hunk 10 above)
Comment (hhamon) [inline src/Symfony/Component/DependencyInjection/Container.php]: yes that's right!

File: src/Symfony/Component/DependencyInjection/Container.php (hunk 11 above)
Comment (hhamon) [inline src/Symfony/Component/DependencyInjection/Container.php]: It may be confusing for people thinking they are overriding a private service but they'll get a new one in the end...

Comment (Ener-Getick) [inline src/Symfony/Component/DependencyInjection/Container.php]: That's a bad practice imo. Why would someone override a service if it is not synthetic ?

Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Container.php]: I agree with @fabpot here, it's the same for private properties on objects btw: they exists from the inside but not from the outside so that one can use the prop name without any collision.

Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Container.php]: Which makes me wonder: it `set` really useful on containers? Shouldn't we deprecate it?

Comment (Ener-Getick) [inline src/Symfony/Component/DependencyInjection/Container.php]: @nicolas-grekas it's useful for synthetic services at least.

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php [hunk 12]
```diff
@@ -642,7 +642,7 @@ private function addService($id, $definition)
      *$lazyInitializationDoc
      * $return
      */
-    {$visibility} function {$methodName}($lazyInitialization)
+    protected function {$methodName}($args)
```
Comment (Ener-Getick) [inline src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: imo you should not change this method visibility in this PR as this is what breaks the tests and due to how the tests are executed they won't pass until the fix is merged, so we could forget bugs...

**Comment by Ener-Getick** [comment]
The tests are missing for the ``ContainerBuilder`` :)

File: src/Symfony/Component/DependencyInjection/Container.php (hunk 11 above)
Comment (hhamon) [inline src/Symfony/Component/DependencyInjection/Container.php]: I'm wondering if we should not have two different Container classes in Symfony 4.0. Something like a `SimpleServiceContainer` class that doesn't support private services and that justs allows to register and resolve definitions. Then the real `Container` class that allows to support private service and dumped containers when they're built from a `ContainerBuilder`.

File: src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php (hunk 12 above)
Comment (hhamon) [inline src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php]: Yes I think i'm gonna revert this change and make it in another separate PR.

**Comment by hhamon** [comment]
Thanks @Ener-Getick! I'm going to add some ;)

File: src/Symfony/Component/DependencyInjection/Container.php (hunk 11 above)
Comment (xabbuh) [inline src/Symfony/Component/DependencyInjection/Container.php]: I suggested that before when we deprecated calling get() before the container is compiled: https://github.com/symfony/symfony/pull/18728#diff-for-comment-62531184

So yes I would say let's deprecate the method entirely for services that are not synthetic.

Comment (stof) [inline src/Symfony/Component/DependencyInjection/Container.php]: ``set('doctrine.orm.entity_manager', null)`` is used by the Doctrine manager registry (in the bridge) to reset the entity manager service (due to the fact that the entity manager becomes unusable after an exception during a flush, because the unit of work is know to be in a broken state).
So we should still at least support this use case.
Allowing to reset an existing instance of the ORM cannot happen without a major change in the ORM (which I will ask to be done in Doctrine 3, but this is not planned to happen soon), so we need to be able to reset the service, to be able to use the ORM in long running scripts like a RabbitMQ consumer (if we cannot reset the service, the only solution would be to kill the process)

**Comment by hhamon** [comment]
I guess it's now ready for another round of reviews if tests pass.

**Comment by nicolas-grekas** [comment]
:+1: (fabbot fails on a fixture, unrelated)

File: src/Symfony/Component/DependencyInjection/Container.php (hunk 11 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Container.php]: See #19192

File: UPGRADE-4.0.md [hunk 13]
```diff
@@ -18,6 +18,15 @@ DependencyInjection
 
  * Using unsupported options to configure service aliases raises an exception.
 
+ * Setting or unsetting a private service with the `Container::set()` method is
+   no longer supported. Only public services can be set or unset.
+
+ * Checking the existence of a private service with the `Container::has()`
+   method is no longer supported.
```
Comment (fabpot) [inline UPGRADE-4.0.md]: It's not that it is not supported anymore but more that it's going to return false.

**Comment by fabpot** [comment]
Apart from my minor comment, 👍

**Comment by hhamon** [comment]
@fabpot fixed! CI is running ;)

**Comment by hhamon** [comment]
Good to go!

**Comment by fabpot** [comment]
Thank you @hhamon.
